mlx/backend/metal/utils.h

```diff
@@ -162,6 +163,48 @@ collapse_contiguous_dims(Arrays... xs) {
       std::vector<array>{std::forward<Arrays>(xs)...});
 }
 
+#ifdef MLX_METAL_DEBUG
+
+NS::String* _debug_make_string(std::ostringstream& os) {
+  return NS::String::string(os.str().c_str(), NS::UTF8StringEncoding);
+}
+
+void _debug_set_stream_queue_label(MTL::CommandQueue* queue, int index) {
+  std::ostringstream label;
+  label << "Stream " << index;
+  queue->setLabel(_debug_make_string(label));
+}
+
+void _debug_push_group(Primitive& primitive) {
+  std::ostringstream group;
+  primitive.print(group);
+
+  auto stream = primitive.stream();
+  auto& device = metal::device(stream.device);
+  auto compute_encoder = device.get_command_encoder(stream.index);
+  compute_encoder->pushDebugGroup(_debug_make_string(group));
+}
+
+void _debug_pop_group(Primitive& primitive) {
+  auto stream = primitive.stream();
+  auto& device = metal::device(stream.device);
+  auto compute_encoder = device.get_command_encoder(stream.index);
+  compute_encoder->popDebugGroup();
+}
+
+#define mlx_metal_debug_set_stream_queue_label(queue, index) \
+  _debug_set_stream_queue_label(queue, index)
+#define mlx_metal_debug_push_group(primitive) _debug_push_group(primitive)
+#define mlx_metal_debug_pop_group(primitive) _debug_pop_group(primitive)
+
+#else
+
+#define mlx_metal_debug_set_stream_queue_label(queue, index)
+#define mlx_metal_debug_push_group(primitive)
+#define mlx_metal_debug_pop_group(primitive)
+
+#endif
```

Comment:
Why not just make these regular functions?

Also a bit of a nit but the names are especially long. I would remove `mlx_` since they are already in the `mlx` namespace. seems superflous.

Reply:
I chose macros to ensure zero debug code execution when `MLX_METAL_DEBUG` is `ON`, without having to also guard call sites with `MLX_METAL_DEBUG`. Gave them `mlx_metal`-prefixed names because they're macros. 

I don't have any strong objections to functions like so:

```cpp
void debug_set_stream_queue_label(MTL::CommandQueue* queue, int index) {
#ifdef MLX_METAL_DEBUG
  std::ostringstream label;
  label << "Stream " << index;
  queue->setLabel(_debug_make_string(label));
#endif
}
```

Is that what you were thinking?

Reply:
Pretty much, this way its consistent with the other debug functions

Reply:
Promoted to proper functions.